cleverhans/feat_advs.py

```diff
@@ -0,0 +1,290 @@
+    :param preds: the model's output tensor
+    :param y: (optional) A placeholder for the model labels. If targeted
+              is true, then provide the target label. Otherwise, only provide
+              this parameter if you'd like to use true labels when crafting
+              adversarial samples. Otherwise, model predictions are used as
+              labels to avoid the "label leaking" effect (explained in this
+              paper: https://arxiv.org/abs/1611.01236). Default is None.
+              Labels should be one-hot-encoded.
+    :param eps: the epsilon (input variation parameter)
+    :param ord: (optional) Order of the norm (mimics NumPy).
+                Possible values: np.inf, 1 or 2.
+    :param clip_min: Minimum float value for adversarial example components
+    :param clip_max: Maximum float value for adversarial example components
+    :param targeted: Is the attack targeted or untargeted? Untargeted, the
+                     default, will try to make the label incorrect. Targeted
+                     will instead try to move in the direction of being more
+                     like y.
+    :return: a tensor for the adversarial example
+    """
+
+    # feat.shape = (batch, c) or (batch, w, h, c)
+    axis = range(1, len(s_feat.shape))
+
+    # Compute loss (targeted)
+    loss = tf.reduce_sum(tf.square(s_feat - g_feat), axis)
+
+    if targeted:
+        loss = - loss
+
+    # Define gradient of loss wrt input
+    grad, = tf.gradients(loss, x)
+
+    if ord == np.inf:
+        # Take sign of gradient
+        signed_grad = tf.sign(grad)
+    elif ord == 1:
+        reduc_ind = list(xrange(1, len(x.get_shape())))
+        signed_grad = grad / tf.reduce_sum(tf.abs(grad),
+                                           reduction_indices=reduc_ind,
+                                           keep_dims=True)
+    elif ord == 2:
+        reduc_ind = list(xrange(1, len(x.get_shape())))
+        signed_grad = grad / tf.sqrt(tf.reduce_sum(tf.square(grad),
+                                                   reduction_indices=reduc_ind,
+                                                   keep_dims=True))
+    else:
+        raise NotImplementedError("Only L-inf, L1 and L2 norms are "
+                                  "currently implemented.")
+
+    # Multiply by constant epsilon
+    scaled_signed_grad = eps * signed_grad
+
+    # Add perturbation to original example to obtain adversarial example
+    adv_x = tf.stop_gradient(x + scaled_signed_grad)
+
+    # If clipping is needed, reset all values outside of [clip_min, clip_max]
+    if (clip_min is not None) and (clip_max is not None):
+        adv_x = tf.clip_by_value(adv_x, clip_min, clip_max)
+
+    return adv_x
+
+
+class FastFeatureGradient(Attack):
+
+    """
+    This is similar to FGSM but done for the internal representations.
+    """
+
+    def __init__(self, model, back='tf', sess=None):
+        """
+        Create a FastGradientMethod instance.
```

Comment:
This should be `Create a FastFeatureGradient instance`, right?

Reply:
Fixed in c80b238. Thank you.